core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -211,16 +211,24 @@ private void losePrimacy() throws Exception {
 
   @Override
   public void stop() throws Exception {
-    LOG.info("Stopping...");
-    mRunning = false;
-    super.stop();
-    if (mLeaderSelector != null) {
-      mLeaderSelector.stop();
+    synchronized (mIsStopped) {
+      if (mIsStopped.get()) {
+        return;
+      }
+      LOG.info("Stopping...");
+      mRunning = false;
+      stopCommonHAAndNonHAServices();
+      if (mLeaderSelector != null) {
+        mLeaderSelector.stop();
+      }
+      mIsStopped.set(true);
+      LOG.info("Stopped.");
     }
   }
 
   /**
-   * @return whether the master is running
+   * @return {@code true} when {@link #start()} has been called and {@link #stop()} has not yet
+   * been called, {@code false} otherwise
    */
   boolean isRunning() {
```

Comment:
They are not, as `mRunning` is set to `false` at the beginning of `stop()`, while `mIsStopped` is set to true at the end of `stop()`. This distinction is important because of the problems with the shutdown sequence we've experienced. `mIsStopped` tells us that all resources have been cleaned up, whereas `mIsRunning` does not.